src/CMakeLists.txt

```diff
@@ -117,14 +118,14 @@ if (NOT("${CHECK_OLEAN_VERSION}" MATCHES "ON"))
   set(LEAN_EXTRA_CXX_FLAGS "${LEAN_EXTRA_CXX_FLAGS} -D LEAN_IGNORE_OLEAN_VERSION")
 endif()
 
-if("${CMAKE_C_COMPILER}" MATCHES "emcc")
-    set(EMSCRIPTEN ON)
+if(${CMAKE_SYSTEM_NAME} MATCHES "Emscripten")
     set(MULTI_THREAD OFF)
-    set(TCMALLOC OFF)
-    set(CFLAGS_EMSCRIPTEN "-Oz -O3")
-    set(LEAN_EXTRA_CXX_FLAGS "${LEAN_EXTRA_CXX_FLAGS} ${CFLAGS_EMSCRIPTEN} -D LEAN_EMSCRIPTEN")
-    set(LEAN_EXTRA_LINKER_FLAGS "${LEAN_EXTRA_LINKER_FLAGS} --memory-init-file 0 --llvm-lto 1 -s ALLOW_MEMORY_GROWTH=1 -s DISABLE_EXCEPTION_CATCHING=0 ${CFLAGS_EMSCRIPTEN}")
-    set(LEAN_JS_LIBRARY "${CMAKE_CURRENT_SOURCE_DIR}" CACHE STRING "location of olean files for lean.js")
+    # From https://emscripten.org/docs/compiling/WebAssembly.html#backends:
+    # > The simple and safe thing is to pass all -s flags at both compile and link time.
+    set(EMSCRIPTEN_SETTINGS "-s ALLOW_MEMORY_GROWTH=1 -s DISABLE_EXCEPTION_CATCHING=0 -s MAIN_MODULE=1 -fexceptions")
+    set(LEAN_EXTRA_CXX_FLAGS "${LEAN_EXTRA_CXX_FLAGS} -D LEAN_EMSCRIPTEN ${EMSCRIPTEN_SETTINGS}")
+    set(LEAN_EXTRA_LINKER_FLAGS "${LEAN_EXTRA_LINKER_FLAGS} ${EMSCRIPTEN_SETTINGS}")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${EMSCRIPTEN_SETTINGS}")
```

Comment:
To have the custom optimisation flags set below apply to C as well, this would actually have to be set for all build types (`CMAKE_C_FLAGS_RELEASE` etc), as currently only the C++ receives them. So libInit/Std/Lean are built with default CMake C options.

Reply:
Ah, you are correct. It looks like this worked only by chance so far. For example, if we actually made use of `LEAN_MULTI_THREAD` in a runtime header, the stage 0 server would crash since that flag would not be set. I think for consistency we should build the bootstrap .c files using `leanc` (i.e. respecting `LEANC_EXTRA_FLAGS`). What do you think?
/cc @leodemoura

Reply:
This would need some adapting in `leanc` since it's currently instantiated once from a template into `stage0/bin/leanc` which contains all `-lInit/Std/Lean`, i.e. not what we want. I think the approach of just adding stuff to `CMAKE_C_FLAGS` could be okay.

Reply:
Okay, I implemented this in the last commit. Now all C code is supposed to go through `leanc`. I added `LEANC_EXTRA_C_FLAGS` for flags that should also be given to compilers if enabled in the build, for example the previously hardcoded `-D LEAN_MULTI_THREAD`.

Reply:
If the flags are used for both compiling and linking, why not keep the old name?

Reply:
They were not included in the compiling flags, but I'm happy to revert to the old name and pass it everywhere. I think having basically the same set of flags in both, with potentially a few extras for the linker, is the simplest and best approach -- if it works.